### ISSUE TITLE
Flag to type `Promise.catch` variables as `unknown`

### ISSUE THREAD

**Issue by yoav-lavi** [microsoft/TypeScript#45602]
# Suggestion

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Following  #41016 a flag was added to consider `catch` variables as `unknown`.
For consistency and since they're similar cases, I suggest also adding a flag that does this for `.catch` callbacks on promises.

See also: [eslint-plugin-etc/no-implicit-any-catch](https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-implicit-any-catch.md)

**Comment by andrewbranch** [comment]
The type for `.catch(...)` callback parameters is defined in lib files, so we’d have to expose a new type that acts like `unknown` or `any` depending on compilation settings... and I’m not sure we’d want that type escaping into the wild 😬. This is a compelling use case, though.

**Comment by Cellule** [comment]
You can easily achieve this by overloading the signature of `Promise.catch` in your project

```ts
// Promise.d.ts
interface Promise<T> {
  /**
   * Attaches a callback for only the rejection of the Promise.
   * @param onrejected The callback to execute when the Promise is rejected.
   * @returns A Promise for the completion of the callback.
   */
  catch<TResult = never>(
    onrejected?: ((reason: unknown) => TResult | PromiseLike<TResult>) | undefined | null,
  ): Promise<T | TResult>;
}
```

**Comment by fregante** [comment]
@Cellule thanks for that! Unfortunately it doesn't cover `allSettled` though and this piece of code doesn't work:

```ts
interface PromiseRejectedResult {
  status: "rejected";
  // @ts-expect-error TS2717 even if overridden, it doesn't work
  reason: unknown;
}
```

https://www.typescriptlang.org/play?target=7#code/IYZwngdgxgBAZgV2gFwJYHsIwG4FMBOYehEqUuAFAJQwDeAUDEzFJiMjAA77oC2qIXCBgBeGMADuwVBwAKPfoIB0wADaqAyrmTJVuACYUA2gEYAulUbM46fDAqsI7LgoG4Y6OC75uQNBszMqF4U3D7K7MDICMIicTAARPi4AFa4UMgGCf5WgUxhirhKyaCYucwAvuVVFUA

<img width="653" alt="Screen Shot 9" src="https://user-images.githubusercontent.com/1402241/144802743-783d2f4f-0733-4045-8437-eee996ee9354.png">

**Comment by emartinfigma** [comment]
> You can easily achieve this by overloading the signature of Promise.catch in your project

I tried this, but it seems like TS just selects the 'any' overload if the signatures don't match.  [Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgApQPYFtgGcIA8AKgHzIDeAUMsgPQBU91N9yAgmGIgBYS7JxkCOABsRAI0QBrZDAxRkGECICeyML2RQIAKwgIwwJYpjrN6bHggA6ZslYABAA5wocLIpDa9BiABNkIk1hMUkEGTAMZAgAD30AV0hkAHdeEDMUCxx8ZDwtXX1IP1sae2QHbTB4qBB+NjRMbJQ5BQ0UBGwnEQhDYwxTNqFRCWkSllpmYTAEbmIAJT54kTBkAF5kEAgAN2gSAAo7T29C-wB+AC5kPb3tOFwlS-iQKRAMZJAASjWyIgXcJZWAB8GpZ8AAZYBSQi-RbLEhfYFPPwQGCgfzIYEgJYiAA0zA+lyyVmIGMCfwBJAA3JQAL6USgdWorGJrDYQZIgprXL6rMjkGkfakxaxTGbXFSXXBgKCgADmPL5AspNFotGiMSchUu0EwUBxyClMpAso2GBWTxebxA9KAA)

**Comment by thw0rted** [comment]
I'd really like to see this implemented, and I think it's a shame it wasn't considered for release alongside `useUnknownInCatchVariables`.  Before the flag existed, you could always freely convert between

* `p.then(good, bad)`
* `p.then(good).catch(bad)`, and
*  `try { result = await p; good(result) } catch (err) { bad(err); }`  (in `async` context)

Now, if `bad` doesn't expect `unknown` it will fail, but only in the last of these forms.  I want a flag that says "use `unknown` type for all errors", including try/catch and Promise rejection -- which are the same thing, under async/await!

**Comment by ExE-Boss** [comment]
@thw0rted
> - `p.then(good, bad)`
> - `p.then(good).catch(bad)`

Are not the same, since if `good` throws an error, only the latter form will catch it.

---

`p => p.then(good, bad)` is actually equivalent to:
```js
async p => {
	let result;
	try {
		result = await p;
	} catch (err) {
		return bad(err);
	}

	return good(result);
}
```

**Comment by thw0rted** [comment]
That's correct, but I didn't mean that those forms were identical, I was only referring to the parameter type of the `bad` function. The point is that the then-callback gets `any` while the catch-block invocation gets `unknown`.

**Comment by kirkwaiblinger** [comment]
I've been working on a PR for this rule. One question I have for the maintainers is: do we want to use the type checker for this rule?

Without it, I have code passing every test case I can think of where there is a _function literal_ in the `.catch` callback. For example, the following all flag:
```ts
Promise.reject().catch(x => f(x));
Promise.reject().catch(function(x) { return f(x)});
Promise.reject().catch((...x: [string]) => f(x));
```
However, the following types of cases present some trouble without the type checker:
```ts
declare const yoloHandler: (x: number) => void;
Promise.reject().catch(yoloHandler);

function yoloHandler(x: number): void { }
Promise.reject().catch(yoloHandler);

const cursedCode =  (...args) => Promise.reject().catch(...args);
cursedCode((x => 'should flag with appropriate mental gymnastics');
```
At least the first two examples above would be surgically flaggable with the type checker. Or, alternately, I could just write a rule that requires a function literal, regardless of the type passed in, i.e. flagging on the following, even though it does actually have the correct signature
```ts
function handlerThatUsesUnknown(x: unknown): void { }
Promise.reject().catch(handlerThatUsesUnknown);
```

Technically that's a false positive.... but always using an arrow function is pretty idiomatic anyway, so I, personally, wouldn't feel too bad about if it flagged there and even came with a suggestion to transform it to 
```ts
Promise.reject().catch((err: unknown) => handlerThatUsesUnknown(err));
```

Note that the other tradeoff is that without the type checker, we'll be inspecting _any_ `.catch` method call, regardless of whether its type is actually Promise.prototype.catch or just some random method unfortunately named `.catch` on any old object.

So we'll get the false positive 
```ts
const hasCatch = {
    catch(f: (x: string) => string)): void {
         console.log(f('catch'));
    }
}
hasCatch.catch((x: string) => x);
```

My intuition here is that it's probably reasonable to take the compromise of forcing function literals and avoid the type checker. But I don't have a good sense of the balance between edge-case correctness and type-checker aversion, so, I'd love for a maintainer to chime in! Thanks :)

**Comment by NotWoods** [comment]
Could the approach used to fix iterator types in #58243 be used to fix `promise.catch` as well?

Similiar to the `BuiltinIteratorReturn` type, perhaps a `PromiseCatchErrorType` can be used that would be either `any` or `unknown` depending on a compiler flag.

```ts
interface Promise<T> {
    catch<TResult = never>(onrejected?: ((reason: PromiseCatchErrorType) => TResult | PromiseLike<TResult>) | undefined | null): Promise<T | TResult>;
}
```

**Comment by jfxio** [comment]
Probably obvious, but slight extension to @Cellule's Promise type to also infer `unknown` for the `onrejected` `then()` parameter. This covers both `bad` uses in @thw0rted's reply.

```typescript
// Promise.d.ts
interface Promise<T> {
  then<TFul, TRej = never>(
    onfulfilled?: ((value: T) => TFul | PromiseLike<TFul>) | null,
    onrejected?: ((reason: unknown) => TRej | PromiseLike<TRej>) | null,
  ): Promise<TFul | TRej>

  catch<TRej = never>(
    onrejected?: ((reason: unknown) => TRej | PromiseLike<TRej>) | null,
  ): Promise<T | TRej>
}
```

_Node 20.12, TS ^5.7.2 w/ target & lib = ESNext_
